openlibrary/catalog/add_book/match.py

```diff
@@ -53,13 +47,434 @@ def editions_match(candidate, existing):
     ):
         if existing.get(f):
             rec2[f] = existing[f]
+    # Transfer authors as Dicts str: str
     if existing.authors:
         rec2['authors'] = []
-        for a in existing.authors:
-            while a.type.key == '/type/redirect':
-                a = web.ctx.site.get(a.location)
-            if a.type.key == '/type/author':
-                assert a['name']
-                rec2['authors'].append({'name': a['name'], 'db_name': db_name(a)})
+    for a in existing.authors:
+        while a.type.key == '/type/redirect':
+            a = web.ctx.site.get(a.location)
+        if a.type.key == '/type/author':
+            author = {'name': a['name']}
+            if birth := a.get('birth_date'):
+                author['birth_date'] = birth
+            if death := a.get('death_date'):
+                author['death_date'] = death
+            rec2['authors'].append(author)
+    return threshold_match(rec, rec2, threshold)
+
+
+def normalize(s: str) -> str:
+    """
+    Normalizes a title for matching purposes, not display,
+    by lowercasing, unicode -> NFC,
+    stripping extra whitespace and punctuation, and replacing ampersands.
+    """
+
+    if isinstance(s, str):
+        # LATIN SMALL LETTER L WITH STROKE' (U+0142) -> 'l'
+        s = unicodedata.normalize('NFC', s.replace('\u0142', 'l'))
+    s = s.replace(' & ', ' and ')
+    # remove {mlrhring} and friends
```

Comment:
Remove these comments and un-used code relating to MARC mnemonics. If this needs to be done, it should be in our MARC processing code (maybe it is already) -- this code is all about matching algorithms. The basic content of the title strings should be already extracted at this point to edition import records, form any source.

Reply:
And, of course, replacing ampersands with the word "and" won't work for any non-English language. I haven't thought it through completely, but might be better to work in the other direction and then you could replace all of " and ", " und ", " et ", etc with " & ".

Reply:
There are a number of tests that rely on this ampersand replacement functionality working the way it does currently, I don't want to change it as part of this refactor tidy up, although I agree it seems odd and possibly ineffectual.

It looks like we will not match "A und B" with "A & B", but we will match "A and B" to "A & B". I'd rather figure out coherent matching strategies and use cases and implement those than keep patching old code because it exists.